src/common/opt_partition_builder.h

```diff
@@ -0,0 +1,551 @@
+      return scolumns8;
+  }
+
+  std::vector<const SparseColumn<uint16_t>*>& GetSparseColumnsRefImpl(const uint16_t* dummy) {
+      return scolumns16;
+  }
+
+  std::vector<const SparseColumn<uint32_t>*>& GetSparseColumnsRefImpl(const uint32_t* dummy) {
+      return scolumns32;
+  }
+
+  template <typename BinIdxType>
+  void Init(GHistIndexMatrix const& gmat, const ColumnMatrix& column_matrix,
+            const RegTree* p_tree_local, size_t nthreads, size_t max_depth,
+            size_t n_rows, bool is_lossguide) {
+    gmat_n_rows = gmat.row_ptr.size() - 1;
+    p_tree = p_tree_local;
+    if ((states.size() == 0 && !gmat.IsDense()) ||
+        (data_hash != reinterpret_cast<const uint8_t*>(column_matrix.GetIndexData())
+        && !gmat.IsDense())) {
+      states.clear();
+      default_flags.clear();
+      states.resize(nthreads);
+      default_flags.resize(nthreads);
+      GetColumnsRef<BinIdxType>().resize(column_matrix.GetNumFeature());
+      GetDenseColumnsRef<BinIdxType>().resize(column_matrix.GetNumFeature());
+      GetSparseColumnsRef<BinIdxType>().resize(column_matrix.GetNumFeature());
+      for (size_t fid = 0; fid < column_matrix.GetNumFeature(); ++fid) {
+        GetColumnsRef<BinIdxType>()[fid] = column_matrix.GetColumn<BinIdxType, true>(fid);
+        GetDenseColumnsRef<BinIdxType>()[fid] = dynamic_cast<const DenseColumn<BinIdxType, true>*>(
+                                                GetColumnsRef<BinIdxType>()[fid].get());
+        GetSparseColumnsRef<BinIdxType>()[fid] =
+          dynamic_cast<const SparseColumn<BinIdxType>*>(GetColumnsRef<BinIdxType>()[fid].get());
+      }
+      for (size_t tid = 0; tid < nthreads; ++tid) {
+        states[tid].resize(1 << (max_depth + 1), 0);
+        default_flags[tid].resize(1 << (max_depth + 1), false);
+      }
+    }
+    data_hash = reinterpret_cast<const uint8_t*>(column_matrix.GetIndexData());
+    n_threads = nthreads;
+    this->max_depth = max_depth;
+    vec_rows.resize(nthreads);
+    if (is_lossguide) {
+      partitions.resize(1 << (max_depth + 2));
+      vec_rows_remain.resize(nthreads);
+    } else {
+      threads_nodes_count.clear();
+      threads_nodes_count.resize(n_threads);
+    }
+    if (vec_rows[0].size() == 0) {
+      size_t chunck_size = n_rows / nthreads + !!(n_rows % nthreads);
+    #pragma omp parallel num_threads(n_threads)
+      {
+        size_t tid = omp_get_thread_num();
+        if (vec_rows[tid].size() == 0) {
+          vec_rows[tid].resize(chunck_size + 2, 0);
+          if (is_lossguide) {
+            vec_rows_remain[tid].resize(chunck_size + 2, 0);
+          }
+        }
+      }
+    }
+    threads_addr.clear();
+    threads_id_for_nodes.clear();
+    node_id_for_threads.clear();
+    nodes_count.clear();
+  }
+
+  template<typename BinIdxType, bool is_loss_guided, bool all_dense = true>
+  void CommonPartition(size_t tid, const size_t row_indices_begin,
```

Comment:
Have you tried a complexity analysis tool?

Reply:
No. Hope in separate PR's it would be easy to review. At least I can split this kernel into two independent parts: sparse and dense